pilot/pkg/xds/debug.go

```diff
@@ -330,15 +330,26 @@ func (s *DiscoveryServer) distributedVersions(w http.ResponseWriter, req *http.R
 			con.proxy.RLock()
 
 			if con.proxy != nil && (proxyNamespace == "" || proxyNamespace == con.proxy.ConfigNamespace) {
+				// if there is no value from last nonce,
+				// "-1" should be set for identification of no last nonce
+				sclusterType := s.StatusReporter.QueryLastNonce(con.ConID, v3.ClusterType)
+				if len(sclusterType) == 0 {
+					sclusterType = "-1"
+				}
+				slistenerType := s.StatusReporter.QueryLastNonce(con.ConID, v3.ListenerType)
+				if len(slistenerType) == 0 {
+					slistenerType = "-1"
+				}
+				srouteType := s.StatusReporter.QueryLastNonce(con.ConID, v3.RouteType)
+				if len(srouteType) == 0 {
+					srouteType = "-1"
+				}
 				// read nonces from our statusreporter to allow for skipped nonces, etc.
 				results = append(results, SyncedVersions{
-					ProxyID: con.proxy.ID,
-					ClusterVersion: s.getResourceVersion(s.StatusReporter.QueryLastNonce(con.ConID, v3.ClusterType),
-						resourceID, knownVersions),
-					ListenerVersion: s.getResourceVersion(s.StatusReporter.QueryLastNonce(con.ConID, v3.ListenerType),
-						resourceID, knownVersions),
-					RouteVersion: s.getResourceVersion(s.StatusReporter.QueryLastNonce(con.ConID, v3.RouteType),
-						resourceID, knownVersions),
+					ProxyID:         con.proxy.ID,
+					ClusterVersion:  s.getResourceVersion(sclusterType, resourceID, knownVersions),
```

Comment:
rather than passing -1 to the getResourceVersion function, why not short circuit above and if the nonce len = 0, append a result using -1?

Reply:
@therealmitchconnors According to the code,  the connection ID still can be accessible from connection clients, and any last nonce of `v3.ClusterType`, `v3.ListenerType` and `v3.RouteType` may be empty, I think that passing `-1` can help to identify this situation, rather than ignore them.

Reply:
I like what you're trying to do, but I think this is the wrong place to solve this problem.  the QueryLastNonce() function should not be returning empty.  Can you help me understand what circumstances lead to the Route value being empty there?  I'm not able to reproduce...

Reply:
Hi @therealmitchconnors ,  it's very odd that I have not reproduced the hung bug since I checkout to istio master branch today, and it's similar with this issue https://github.com/istio/istio/issues/27330 . I doubt that someone has fixed it, such as this one: https://github.com/istio/istio/issues/31509#issuecomment-802911104

Reply:
Hi @therealmitchconnors ,  I notice that there were some changes for `https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go` in 2 days, so this bug may be fixed,  I think this issue should be closed.  However, I think there should be some changes to optimize these code, such as:    `pilot/pkg/xds/debug.go` and `istioctl/cmd/wait.go`,  if necessary, I can submit PR for them. What do you think about it?